<h2 style="color:blue;"> HW06 - Scenario 12  </h2>


### Import data and creating the vectorized Train/Test Split 

In [3]:
#Code Block 1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [13]:
#Code Block 2


#URL = 'https://data6310.file.core.windows.net/data6310hw/PodcastFilm.xlsx?sp=rl&st=2020-11-12T21:45:20Z&se=2022-12-13T21:45:00Z&sv=2019-12-12&sig=u0hS%2FvAC9s8pJUEYHbt7chCxh09M8z3rGVdw3ifBQKc%3D&sr=f'

podcast_film =  pd.ExcelFile('Homework6_Data/PodcastFilm.xlsx') 
print(podcast_film.sheet_names)
df = podcast_film.parse('podcasts', index_col=0)

['podcasts']


In [14]:
#Code Block 3

df.head()

podcast_id   itunes_id  \
97   a00ac2a5a4c20cf44850d76d7e14aab2  1460410612   
128  a00cb03b8f759d95d3ed9d1e12f94302   980276589   
129  a00cb03b8f759d95d3ed9d1e12f94302   980276589   
130  a00cb03b8f759d95d3ed9d1e12f94302   980276589   
131  a00cb03b8f759d95d3ed9d1e12f94302   980276589   

                                 slug  \
97   eana-mokri-the-learnnavi-podcast   
128       to-the-batpoles-batman-1966   
129       to-the-batpoles-batman-1966   
130       to-the-batpoles-batman-1966   
131       to-the-batpoles-batman-1966   

                                            itunes_url  \
97   https://podcasts.apple.com/us/podcast/eana-mok...   
128  https://podcasts.apple.com/us/podcast/to-the-b...   
129  https://podcasts.apple.com/us/podcast/to-the-b...   
130  https://podcasts.apple.com/us/podcast/to-the-b...   
131  https://podcasts.apple.com/us/podcast/to-the-b...   

                                   title category  \
97   Eana Mokri - The LearnNa'vi Podcast  tv-film   
128         To The Batpoles! Batman 1966  tv-film   
129         To The Batpoles! Batman 1966  tv-film   
130         To The Batpoles! Batman 1966  tv-film   
131         To The Batpoles! Batman 1966  tv-film   

                                      review_title  \
97                                    Best Podcast   
128                      Keep Your Bat-radio Tuned   
129    A Must-Listen Pocast for Any Batman '66 Fan   
130  In-depth insights with a side order of fanboy   
131                         Batbest of the Batpods   

                                               content  rating  \
97   This is by far the best podcast for everything...       5   
128  I've nothing to add to the previous reviews.  ...       5   
129  As soon as I bought the Batman series on DVD, ...       5   
130  Tim and Paul not only have an excellent on-air...       5   
131  Since I've come across "To the Batpoles," my u...       5   

                    created_at  
97   2019-05-06T07:50:05-07:00  
128  2017-09-15T10:03:49-07:00  
129  2018-02-02T09:12:58-07:00  
130  2018-10-11T05:05:56-07:00  
131  2019-01-10T19:13:23-07:00

In [16]:
#Code Block 4

def status5(d):
    if d['rating'] == 5:
        return "5-Star"
    else:
        return "4 or less"
df['Status_5'] = df.apply(status5, axis = 1)

In [19]:
#Code Block 5

def status1(d):
    if d['rating'] == 1:
        return "1-Star"
    else:
        return "2 or higher"
df['Status_1'] = df.apply(status1, axis = 1)

In [21]:
#Code Block 6

df.sample(10, random_state = 42)

podcast_id   itunes_id  \
372395  c290a84c4aec6b88dcac43d8a072b9da   573656380   
450862  c8f432ae7626ca496edb283ac2a2a0e3   467892214   
404751  c4f5b1574044226d89415200e5ceda6d  1124616634   
979158  f6dc57e9806e1162d633b56c04c560bf   657408550   
500657  cda7e73db840cb814ad6b31306078d1f   624693800   
562657  d310a842223f03e1a5e7e372b9204c5b   983639868   
582086  d4baff0192b5f556b0c3f2a273e98e93  1345682353   
155812  b00702f0db9be549df4d5d1ea9503431   600020121   
406157  c4fa1227834a55e7f1abd93142b9be38  1195238167   
561057  d2d3371547bd35f451d0b027db2f3a6e   453968761   

                                                    slug  \
372395                                    aerial-america   
450862                                        radiodrome   
404751  take-the-black-podcast-a-game-of-thrones-podcast   
979158  children-shouldnt-play-with-dead-formats-podcast   
500657                                       who-shot-ya   
562657                           game-of-thrones-podcast   
582086                  armchair-expert-with-dax-shepard   
155812                                  mad-about-movies   
406157                                  the-b-scott-show   
561057                               podcast-of-interest   

                                               itunes_url  \
372395  https://podcasts.apple.com/us/podcast/aerial-a...   
450862  https://podcasts.apple.com/us/podcast/radiodro...   
404751  https://podcasts.apple.com/us/podcast/take-the...   
979158                                                NaN   
500657  https://podcasts.apple.com/us/podcast/who-shot...   
562657  https://podcasts.apple.com/us/podcast/game-of-...   
582086  https://podcasts.apple.com/us/podcast/armchair...   
155812  https://podcasts.apple.com/us/podcast/mad-abou...   
406157  https://podcasts.apple.com/us/podcast/the-b-sc...   
561057  https://podcasts.apple.com/us/podcast/podcast-...   

                                                    title category  \
372395                                     Aerial America  tv-film   
450862                                         Radiodrome  tv-film   
404751  Take the Black Podcast, a Game of Thrones Podcast  tv-film   
979158  Children Shouldn't Play With Dead Formats' Pod...  tv-film   
500657                                       Who Shot Ya?  tv-film   
562657                            Game of Thrones Podcast  tv-film   
582086                   Armchair Expert with Dax Shepard  tv-film   
155812                                   Mad About Movies  tv-film   
406157                                  The B. Scott Show  tv-film   
561057                                Podcast Of Interest  tv-film   

                                   review_title  \
372395  Takes 20 minutes to listen to 2 minutes   
450862                                 The best   
404751                           Host Chews Ice   
979158     Travel Back in Time with these guys!   
500657                              I’m hooked!   
562657                                Wonderful   
582086                     All the gold stars!!   
155812                                    Ooof!   
406157                            Amaaaaaazing!   
561057                                  5 stars   

                                                  content  rating  \
372395  Concept great, execution nonexistent. \nThe vi...       1   
450862  The best movie discussion podcast of the kind ...       5   
404751  Sometimes it’s a good show and sometimes the h...       1   
979158  Their obsessive knowledge of dear formats and ...       5   
500657  Started listening recently when Story Break’s ...       5   
562657  I love listening to this every week after watc...       5   
582086  I’ve been a listener since day one and while s...       5   
155812  Just the worst. A bunch of very snarky, preten...       1   
406157  Man... this is one of THE best podcasts I have...       5   
561057  By far, unquestionable the best podcast for Pe...

In [23]:
#Code Block 7

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer

In [24]:
#Code Block 8

y = df['Status_5']

In [27]:
#Code Block 9

X = df['content']

### Train test split


In [30]:
#Code Block 10

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [32]:
#Code Block 11

display(y_train.shape)
y_test.shape

(60342,)

(25861,)

In [34]:
#Code Block 12

count_vectorizer = CountVectorizer(stop_words='english')

In [36]:
#Code Block 13

X_train.head()

843417    The 4 Jabronies episode is my jumping on point...
49011     Andy and Austin provide great insight and fun ...
500727    So my name is Des and I’m actually new to the ...
146366                 Enjoy these guys and a great podcast
764395                               Where is Amanduhhhhhh.
Name: content, dtype: object

In [38]:
#Code Block 14

count_train = count_vectorizer.fit_transform(X_train.values)

In [40]:
#Code Block 15

count_test = count_vectorizer.transform(X_test.values)

In [42]:
#Code Block 16

count_test.shape

(25861, 45674)

<h2 style="color:blue;"> Sentiment Analysis for Predicting 5-Star Reviews   </h2>

In [45]:
#Code Block 17

from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics

In [47]:
#Code Block 18

nb_classifier_status = MultinomialNB()

In [49]:
#Code Block 19

nb_classifier_status.fit(count_train, y_train)

MultinomialNB()

In [51]:
#Code Block 20

pred_5 = nb_classifier_status.predict(count_test)

In [53]:
#Code Block 21

metrics.accuracy_score(y_test, pred_5)

0.9137697691504583

In [55]:
#Code Block 22

pred_5 = pd.DataFrame(pred_5, columns = ['pred_5'])
pred_5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25861 entries, 0 to 25860
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   pred_5  25861 non-null  object
dtypes: object(1)
memory usage: 202.2+ KB


In [57]:
#Code Block 23

y_tab = pd.DataFrame(y_test).copy()
y_tab = y_tab.reset_index()
result = pd.concat([y_tab, pred_5], axis=1, join='inner')
result.head(25)

index   Status_5     pred_5
0    372395  4 or less  4 or less
1    450862     5-Star     5-Star
2    404751  4 or less     5-Star
3    979158     5-Star     5-Star
4    500657     5-Star     5-Star
5    562657     5-Star     5-Star
6    582086     5-Star     5-Star
7    155812  4 or less  4 or less
8    406157     5-Star     5-Star
9    561057     5-Star     5-Star
10   829536     5-Star     5-Star
11    75765     5-Star     5-Star
12   325009     5-Star     5-Star
13   687436     5-Star     5-Star
14   307409     5-Star     5-Star
15   393571  4 or less     5-Star
16    32547     5-Star     5-Star
17  1019631     5-Star     5-Star
18   307529     5-Star     5-Star
19   990058     5-Star     5-Star
20   247508     5-Star     5-Star
21   347263     5-Star     5-Star
22   869699     5-Star     5-Star
23   612688     5-Star     5-Star
24    70107     5-Star     5-Star

### Actual 5-Star Reviews¶

In [60]:
#Code Block 24

result['Status_5'].value_counts()

Status_5
5-Star       21786
4 or less     4075
Name: count, dtype: int64

### Predicted 5-Star Reviews¶

In [63]:
#Code Block 25

result['pred_5'].value_counts()

pred_5
5-Star       22336
4 or less     3525
Name: count, dtype: int64

In [65]:
#Code Block 26

ct_status_5 = pd.crosstab(result['Status_5'], result['pred_5'])
ct_status_5['Actual Total'] = ct_status_5['4 or less'] + ct_status_5['5-Star']
ct_status_5

pred_5     4 or less  5-Star  Actual Total
Status_5                                  
4 or less       2685    1390          4075
5-Star           840   20946         21786

In [67]:
#Code Block 27

df_sum = pd.DataFrame(ct_status_5.sum(axis=0)).T
df_sum

pred_5  4 or less  5-Star  Actual Total
0            3525   22336         25861

In [69]:
#Code Block 28

ct_status_5 = pd.concat([ct_status_5, df_sum], axis=0)
ct_status_5

pred_5     4 or less  5-Star  Actual Total
4 or less       2685    1390          4075
5-Star           840   20946         21786
0               3525   22336         25861

In [71]:
#Code Block 29

ct_status_5 = ct_status_5.rename({0: 'Predicted Total'})
ct_status_5

pred_5           4 or less  5-Star  Actual Total
4 or less             2685    1390          4075
5-Star                 840   20946         21786
Predicted Total       3525   22336         25861

In [73]:
#Code Block 30

ct_status_5['4orLess_Percent'] = round(ct_status_5['4 or less'] / ct_status_5['Actual Total'], 2)
ct_status_5['5-Star_Percent'] = round(ct_status_5['5-Star'] / ct_status_5['Actual Total'], 2)
ct_status_5

pred_5           4 or less  5-Star  Actual Total  4orLess_Percent  \
4 or less             2685    1390          4075             0.66   
5-Star                 840   20946         21786             0.04   
Predicted Total       3525   22336         25861             0.14   

pred_5           5-Star_Percent  
4 or less                  0.34  
5-Star                     0.96  
Predicted Total            0.86

### Analysis of 1-Star Crosstab

- We were able to predict 66% of the reviews (4orLess_Percent column - first row) of the 4 or less
- But, we were only able to predict 96% of the 5-Star reviews (5-Star_Percent column - second row)

<h2 style="color:blue;"> Sentiment Analysis for Predicting 1-Star Reviews   </h2>

In [77]:
#Code Block 31

y = df['Status_1']
y

97         2 or higher
128        2 or higher
129        2 or higher
130        2 or higher
131        2 or higher
              ...     
1070092    2 or higher
1070093    2 or higher
1070094    2 or higher
1070095    2 or higher
1070096    2 or higher
Name: Status_1, Length: 86203, dtype: object

In [79]:
#Code Block 32

X = df['content']
X

97         This is by far the best podcast for everything...
128        I've nothing to add to the previous reviews.  ...
129        As soon as I bought the Batman series on DVD, ...
130        Tim and Paul not only have an excellent on-air...
131        Since I've come across "To the Batpoles," my u...
                                 ...                        
1070092    Teddy and the empress continuously make one of...
1070093    Thanks for keeping us sane in this Q time Tedd...
1070094    It’s like they are reading my mind! Love it. L...
1070095    Jokes abound. Laughs a plenty. These hosts are...
1070096    Anything with Teddy and the Empress is an ever...
Name: content, Length: 86203, dtype: object

### Train test split and Vectorization of Status_1

- Create count_train
- Do not need to re-create count_test, since it is the same as Status_5


In [82]:
#Code Block 33

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
count_vectorizer = CountVectorizer(stop_words='english')
count_train = count_vectorizer.fit_transform(X_train.values)
count_test = count_vectorizer.transform(X_test.values)

In [83]:
#Code Block 34

count_vectorizer = CountVectorizer(stop_words='english')
count_train = count_vectorizer.fit_transform(X_train.values)


In [84]:
#Code Block 35

from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics

In [86]:
#Code Block 36

nb_classifier_status = MultinomialNB()

In [90]:
#Code Block 37

nb_classifier_status.fit(count_train, y_train)

MultinomialNB()

In [92]:
#Code Block 38

pred_1 = nb_classifier_status.predict(count_test)

In [94]:
#Code Block 39

metrics.accuracy_score(y_test, pred_1)

0.9435830014307258

In [96]:
#Code Block 40

pred_1 = pd.DataFrame(pred_1, columns = ['pred_1'])
pred_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25861 entries, 0 to 25860
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   pred_1  25861 non-null  object
dtypes: object(1)
memory usage: 202.2+ KB


In [98]:
#Code Block 41

y_tab = pd.DataFrame(y_test).copy()
y_tab = y_tab.reset_index()
result_all = pd.concat([result, y_tab, pred_1], axis=1, join='inner')
result_all = result_all.iloc[:, [0, 1,4,2,5]]
result_all.head(25)

index   Status_5     Status_1     pred_5       pred_1
0    372395  4 or less       1-Star  4 or less  2 or higher
1    450862     5-Star  2 or higher     5-Star  2 or higher
2    404751  4 or less       1-Star     5-Star  2 or higher
3    979158     5-Star  2 or higher     5-Star  2 or higher
4    500657     5-Star  2 or higher     5-Star  2 or higher
5    562657     5-Star  2 or higher     5-Star  2 or higher
6    582086     5-Star  2 or higher     5-Star  2 or higher
7    155812  4 or less       1-Star  4 or less       1-Star
8    406157     5-Star  2 or higher     5-Star  2 or higher
9    561057     5-Star  2 or higher     5-Star  2 or higher
10   829536     5-Star  2 or higher     5-Star  2 or higher
11    75765     5-Star  2 or higher     5-Star  2 or higher
12   325009     5-Star  2 or higher     5-Star  2 or higher
13   687436     5-Star  2 or higher     5-Star  2 or higher
14   307409     5-Star  2 or higher     5-Star  2 or higher
15   393571  4 or less  2 or higher     5-Star  2 or higher
16    32547     5-Star  2 or higher     5-Star  2 or higher
17  1019631     5-Star  2 or higher     5-Star  2 or higher
18   307529     5-Star  2 or higher     5-Star  2 or higher
19   990058     5-Star  2 or higher     5-Star  2 or higher
20   247508     5-Star  2 or higher     5-Star  2 or higher
21   347263     5-Star  2 or higher     5-Star  2 or higher
22   869699     5-Star  2 or higher     5-Star  2 or higher
23   612688     5-Star  2 or higher     5-Star  2 or higher
24    70107     5-Star  2 or higher     5-Star  2 or higher

### Actual 1-Star Reviews

In [101]:
#Code Block 42

result_all['Status_1'].value_counts()

Status_1
2 or higher    24194
1-Star          1667
Name: count, dtype: int64

### Predicted 1-Star Reviews

In [104]:
#Code Block 43

result_all['pred_1'].value_counts()

pred_1
2 or higher    24457
1-Star          1404
Name: count, dtype: int64

### Crosstab of Predicted and Actual
- Rows represent actual 1-Star or 2 or higher.
- Columns are predicted 1-Star or 2 or higher.

In [107]:
#Code Block 44

ct_status_1 = pd.crosstab(result_all['Status_1'], result_all['pred_1'])
ct_status_1['Actual Total'] = ct_status_1['2 or higher'] + ct_status_1['1-Star']
ct_status_1

pred_1       1-Star  2 or higher  Actual Total
Status_1                                      
1-Star          806          861          1667
2 or higher     598        23596         24194

In [109]:
#Code Block 45

df_sum = pd.DataFrame(ct_status_1.sum(axis=0)).T
df_sum

pred_1  1-Star  2 or higher  Actual Total
0         1404        24457         25861

In [111]:
#Code Block 46

ct_status_1 = pd.concat([ct_status_1, df_sum], axis=0)
ct_status_1

pred_1       1-Star  2 or higher  Actual Total
1-Star          806          861          1667
2 or higher     598        23596         24194
0              1404        24457         25861

In [113]:
#Code Block 47

ct_status_1 = ct_status_1.rename({0: 'Predicted Total'})
ct_status_1

pred_1           1-Star  2 or higher  Actual Total
1-Star              806          861          1667
2 or higher         598        23596         24194
Predicted Total    1404        24457         25861

In [115]:
#Code Block 48

ct_status_1['1-Star_Percent'] = round(ct_status_1['1-Star'] / ct_status_1['Actual Total'], 2)
ct_status_1['2orHigher_Percent'] = round(ct_status_1['2 or higher'] / ct_status_1['Actual Total'], 2)
ct_status_1

pred_1           1-Star  2 or higher  Actual Total  1-Star_Percent  \
1-Star              806          861          1667            0.48   
2 or higher         598        23596         24194            0.02   
Predicted Total    1404        24457         25861            0.05   

pred_1           2orHigher_Percent  
1-Star                        0.52  
2 or higher                   0.98  
Predicted Total               0.95

### Analysis of 1-Star Crosstab

- We were able to predict 98% of the reviews (2orHigher_Percent column - second row) of the 2 or higher reviews
- But, we were only able to predict 48% of the 1-Star reviews (1-Star_Percent column - first row)